Rename Symlink paths to improve clarity

omnunum opened the pull request:
I had placed the symlink directory in the rclone directory due to a misunderstanding of both the variable names and the expected outcome.  This PR aims to improve the clarity of the names, as well as simplify and expand the path validation.

omnunum commented on backend/program/symlink.py:
```diff
@@ -27,44 +28,38 @@ class Symlinker():
         if not self.initialized:
-            logger.error("Symlink initialization failed due to invalid configuration.")
-            return
+            raise ValueError("Symlink initialization failed due to invalid configuration.")
```
This change needs discussion

dreulavelle replied:
returning back to parent so that Program handles. If we throw exception from symlink module, does Program throw exception? Hm, needs discussion

omnunum replied:
With the current setup, if the settings fail to initialize we would want to halt the program.  If the failure conditions in settings do not cause the program to halt, inconsistent behavior is to be expected, which can lead the cached pickle data to be in a state that prevents proper functioning when the app is restarted.

Settings should be validated before any actions by the application are taken.

omnunum commented on backend/program/symlink.py:
```diff
@@ -27,44 +28,38 @@ class Symlinker():
+            elif not path.is_absolute():
+                logger.error(f"{name} is not an absolute path: {path}")
+                return False
+            elif not path.is_dir():
```
Not sure why we wouldn't want to validate all entries as dirs

dreulavelle replied:
I felt the same way as well, but Mini was having issues. Need to gather more info from @AyushSehrawat on what kind of issues exactly and **how** it was working for him before, if it even was? Was it playing on plex from the symlink path Mini? Does `/mount` exist? 

It's also possible that I need to understand how `os.symlink()` works and if it requires the path to be a valid path to create a symlink in the OS. This is what I'm thinking happened, is that it doesnt need a valid path to create the link.. but since he plays the symlinks in Plex, that path is valid in plex which works as intended for the `container_path` setting.

dreulavelle commented on backend/program/symlink.py:
```diff
@@ -10,6 +10,7 @@
 class SymlinkConfig(BaseModel):
     host_path: Path
     container_path: Path
+    symlink_path: Path = None
```
why a 3rd path? 

The `container_path` is the path that plex sees. Here's an example, 

![image](https://github.com/dreulavelle/iceberg/assets/5782630/f07cd5ed-de75-4f2b-870c-dc5d958c37f1)

my paths are `/mnt/zurg` for host_path and `/mnt/zurg` for container_path since I set plex volume mount to `/mnt:/mnt`. container_path is the path that Plex see's to be able to read the symlink, and host_path is the rclone path that symlinks get pointed to. Think of it like this,

```
source = container_path
destination = host_path
```

dreulavelle replied:
If we need to change path names to make it more readable/understandable, we can change:

`host_path` -> `rclone_path`
`container_path` -> `symlink_path`

And possibly add a brief description in the frontend about it to users?

dreulavelle replied:
@AyushSehrawat I explain it here

dreulavelle commented on backend/program/symlink.py:
```diff
@@ -80,7 +75,8 @@ def validate(self):
         """Create the initial library folders."""
         try:
-            self.library_path = self.settings.host_path.parent / "library"
+            path = self.settings.symlink_path or self.settings.host_path.parent
```
Why 2 path variations?

omnunum replied:
Because I'm an idiot and didn't realize that's what host path is for.  The naming conventions are not very descriptive of how the paths are used.  In general I think it's presumptuous to assume that users are running plex or this application in a container.  I much prefer your `rclone_path` and `symlink_path` descriptions.

dreulavelle replied:
I'm fine with that, it throws everyone for a loop haha. Go ahead and change the names :D

AyushSehrawat replied:
Let me know to not break frontend 😅

dreulavelle commented on backend/program/symlink.py:
```diff
@@ -27,44 +28,38 @@ class Symlinker():
         symlink_path (str): The path where the symlinks will be created.
     """
-    def __init__(self, _):
+    def __init__(self, *_):
```
Hm, may not understand the need for the *args? Can you explain?

omnunum replied:
I've never seen a requirement to initialize a class with an unused argument before. I understand we have an API we want to stick to, but why require it if we're not using it?  Currently, we have to initialize the class like `Symlink(None)`, which implies that there is some utility to the value.

dreulavelle replied:
If its not being used anywhere, just toss it lol

dreulavelle commented on backend/program/symlink.py:
```diff
@@ -27,44 +28,38 @@ class Symlinker():
-        try:
-            if not host_path.is_dir():
-                logger.error(f"Host path is not a directory or does not exist: {host_path}")
+        for name, path in self.settings.__iter__():
```
I like the simplified approach. I may have went overboard on covering edge cases, to the point it just got messy lol.

dreulavelle commented on dev-requirements.txt:
```diff
@@ -0,0 +1,3 @@
+pytest
+httpx
+black
```
feel free to add these to the main reqs file. My goal is to get CI to run tests before images get built but have been waiting for more tests to get written. Backend code needs to get to a point where it isn't changed as often for certain modules. (core modules)

dreulavelle replied:
Hm.. I'm 50/50 on this one actually. Is this considered best practice? If so, then this is fine.

omnunum replied:
Some folks put the dev requirements in `setup.py`, some make a second file, some use something like poetry to manage them.  I've found poetry to be more annoying than helpful in my day job.  I don't have any strong opinions on the approach so feel free to make a call.

dreulavelle replied:
lol, I thought about checking out poetry. Thanks for saving me the hassle 😂 

We can leave it, thats fine then.

AyushSehrawat replied:
We can just make two requirements text file, see many big projects with it

Also do we use httpx? I remember seeing Requests

omnunum commented:
@dreulavelle Given that I misunderstood the purpose of `host_path` which means that this additional `symlink_path` is wholly unnecessary, what if any parts of this work do you want to merge?

dreulavelle commented:
> what if any parts of this work do you want to merge?

Make the changes to trim it down to just the 2 paths, using the new names of `rclone_path` and `symlink_path`. Change it in the settings file as well. I liked the simplified approach of checking

@AyushSehrawat will need to change frontend to account for settings changes. Also need to add descriptors in frontend to explain.

AyushSehrawat commented:
> > what if any parts of this work do you want to merge?
> 
> Make the changes to trim it down to just the 2 paths, using the new names of `rclone_path` and `symlink_path`. Change it in the settings file as well. I liked the simplified approach of checking the dirs iteratively.
> 
> @AyushSehrawat will need to change frontend to account for settings changes. Also need to add descriptors in frontend to explain.

Let me know the description you want to put, i was asking the same few days ago

dreulavelle commented:
> Zurg appears to be deleting dirs placed in its mount. This means that I need a different location for the symlinks.

Whatttttt?

omnunum commented:
> > Zurg appears to be deleting dirs placed in its mount. This means that I need a different location for the symlinks.
> 
> Whatttttt?

```
rclone-1  | 2024/02/03 18:54:34 DEBUG : /: Mkdir: name="test"
zurg-1    | 2024-02-03T23:54:34.271Z    DEBUG   router  Method not allowed: MKCOL /dav/test/ map[Accept-Encoding:[gzip] Referer:[http://zurg:9999/dav/] User-Agent:[rclone/v1.65.2]]
rclone-1  | 2024/02/03 18:54:34 DEBUG : : Added virtual directory entry vAddDir: "test"
...
rclone-1  | 2024/02/03 18:54:54 DEBUG : test: forgetting directory cache
rclone-1  | 2024/02/03 18:54:54 DEBUG : : Removed virtual directory entry vAddDir: "test"
```
Something goofy was happening where Zurg said that making a new dir wasn't allowed, but rclone made it anyway, but then rclone pruned it (maybe because it was empty?).  Either way that's what spurred me to set the symlink_path to somewhere not in the mount.

dreulavelle commented:
> set the symlink_path to somewhere not in the mount.

It cant be in the mount anyways?
